---
title: "NGS Report"
author: "Institut Curie Bioinformatics Platform"
output: 
  html_document:
    toc: true
    depth: 3
  pdf_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache = FALSE, message=FALSE, warning=FALSE)
```

```{r results="asis", echo=FALSE}
cat("
<style>
h1 {
  color: orange;
  text-align: center;
}
h2 {
  color: blue;
  border-bottom: 8px solid;
}
</style>
")
```

```{r config, include=FALSE}
stopifnot(require(ggplot2))
stopifnot(require(reshape2))
stopifnot(require(plotly))
stopifnot(require(dplyr))
stopifnot(require(magrittr))
stopifnot(require(tidyr))
stopifnot(require(DT))
stopifnot(require(rtracklayer))
stopifnot(require(RColorBrewer))

```

```{r arguments, include=FALSE}
args <- commandArgs()
print(args)
```

```{r function}

plotly_margins <- function(ggplotly_build){
  
  ggplotly_build$layout$margin$b <- ggplotly_build$layout$margin$b + 50
  ggplotly_build$layout$margin$l <- ggplotly_build$layout$margin$l + 50
  
  return(ggplotly_build)
}

fastqc_url <- function(input_path){
  fastqc.dirs <- list.files(input_path, pattern="fastqc$", recursive=TRUE, full.names=TRUE, include.dirs=TRUE)
  if (length(fastqc.dirs)>0){
     sapply (fastqc.dirs, function(dir){        
     	    reports <- list.files(dir, pattern="fastqc.html", full.names=TRUE)
    	    urls=""
      	    for (file in reports){
      	    	urls<-paste0(urls, paste0("<a href='", file, "' target='_blank'>",basename(file),"</a>"), sep="<br>")
    	    }
            urls
      }) 
  }else{
  return(NA) 
 }     
}

inactivate_hover <- function(x){ 
   for (i in 1:length(ggly_b$x$data)){
       x$x$data[[i]]$hoverinfo <- "none"
   }
   x
}

```

| Institut Curie | Report generation date: `r format(Sys.Date(), '%Y-%m-%d')` |
| :------------- | -------------: |
| Report generated by Institut Curie bioinformatic platform | u900.pf.ngs.dm@curie.fr |

#RNA-SEQ ANALYSIS

```{r statsFile, results="asis", eval=TRUE}
statsIn <- list.files(file.path(input_path, "report"), pattern=".fullStatFile.txt", full.names=TRUE)
stopifnot(length(statsIn)==1)
statsXLS <- list.files(file.path(input_path, "report"), pattern=".fullStatFile.xls", full.names=TRUE)
if(length(statsIn)==0) stop("fullStatFile not found !")
statsFile <- read.csv(statsIn, row.names=NULL)

## Format colnames to ease data access
colnames(statsFile) <- c("inputID","sampleID","n_cluster", "n_reads", "n_rrna", "n_non_rrna", "n_aligned", "n_u_aligned", "n_m_aligned", "n_dup", "n_exons", "n_introns", "n_inter")

## Calculate %
statsFile$p_rrna <- paste0(round(statsFile$n_rrna / statsFile$n_reads * 100, 2),"%")
statsFile$p_aligned <- paste0(round(statsFile$n_aligned / statsFile$n_non_rrna * 100, 2),"%")
statsFile$p_u_aligned <- paste0(round(statsFile$n_u_aligned / statsFile$n_non_rrna * 100, 2),"%")
statsFile$p_m_aligned <- paste0(round(statsFile$n_m_aligned / statsFile$n_non_rrna * 100, 2),"%")
statsFile$p_dup <- paste0(round(statsFile$n_dup / statsFile$n_non_rrna * 100, 2),"%")
statsFile$p_exons <- paste0(round(statsFile$n_exons / statsFile$n_u_aligned * 100, 2),"%")
statsFile$p_introns <- paste0(round(statsFile$n_introns / statsFile$n_u_aligned * 100, 2),"%")
statsFile$p_inter <- paste0(round(statsFile$n_inter / statsFile$n_u_aligned * 100, 2),"%")


## FastQC
statsFile$fastqc <- fastqc_url(input_path)
```

##Metadata

```{r meta}
if (file.exists(meta)){
   dmeta <- read.csv(meta, sep="\t", header=TRUE)
   datatable(format(dmeta, big.mark=",",scientific=FALSE), colnames = colnames(dmeta), 
   rownames = FALSE, escape = FALSE, filter = 'top', caption = 'Metadata Information')
}

```

##Data processing

| <font size="5">Species: `r species`</font> | <font size="5">Build: `r build`</font> |
| :------------- | -------------: |
| <font size="5">`r paste("<a href='", statsXLS, "'>Stats file</a>")`</font> | <font size="5">`r paste("<a href='", file.path("./config.txt'"), "'>Config file</a>")`</font> |


<!--| <font size="5">`r paste("<a href='", file.path("./general_graph.svg'"), "target='_blank'>Rule graph</a>")`</font> | <font size="5">`r paste("<a href='", file.path("./sample_graph.svg'"), "target='_blank'>Sample graph</a>")`</font> |-->


```{r table1, eval=TRUE}
datatable(format(statsFile[,c('inputID', 'sampleID', 'n_cluster', 'n_reads', 'n_rrna', 'p_rrna', 'fastqc')], big.mark=",",scientific=FALSE), colnames = c('Sample', 'Biosample', 'Number of clusters', 'Number of total reads', 'Number of rRNA reads', 'Percent of rRNA reads', 'QC reports'), rownames = FALSE, escape = FALSE, filter = 'top', caption = 'Raw Sequencing Reads Information', options = list(columnDefs = list(list(className = 'dt-center', targets = 2:5))))

```
<a href="#">Back to top</a>


##Mapping Statistics

```{r table2, eval=TRUE}

datatable(format(statsFile[,c('inputID', 'sampleID', 'n_non_rrna', 'n_aligned', 'p_aligned', 'n_u_aligned', 'p_u_aligned', 'n_m_aligned', 'p_m_aligned', 'n_dup', 'p_dup')], big.mark=",",scientific=FALSE), colnames =c('Sample', 'Biosample', 'Number of QC-passed reads', 'Number of mapped reads', 'Percent of mapped reads', 'Number of unique reads', 'Percent of unique hits', 'Number of reads mapped to multiple loci', 'Percent of multiple alignments', 'Number of duplicates', 'Percent of duplicates'), rownames = FALSE, filter = 'top', caption = 'Mapping results', options = list(columnDefs = list(list(className = 'dt-center', targets = 2:5))))

```

```{r mappedReadsPlot, fig.width=10, fig.height=7, echo=FALSE, eval=TRUE}
verbosetype <- c("Aligned reads (unique-hits)", "Aligned reads (multi-hits)", "rRNA reads", "Unmapped reads")
names(verbosetype) <- c("uniqueHits","multiHits","rRNA","unmapped")

dataPlot <- statsFile %>%
	select(., inputID, n_reads, n_u_aligned, n_m_aligned, n_rrna) %>% 
	set_colnames(., c("sample", "totalReads", "uniqueHits", "multiHits", "rRNA")) %>% 
	mutate(., unmapped=(totalReads-rRNA-uniqueHits-multiHits)) %>%  
  gather(., key="type", value="n_read", -sample, -totalReads)

## update labels
dataPlot$type=verbosetype[dataPlot$type]
dataPlot$type <- factor(dataPlot$type, levels=c("Aligned reads (unique-hits)", "Aligned reads (multi-hits)", "rRNA reads", "Unmapped reads"))

## Calculate percent
p_unmapped <- paste0(round(dataPlot[which(dataPlot$type=="Unmapped reads"),"n_read"] /     sapply(unique(dataPlot$sample),function(i){sum(dataPlot[which(dataPlot$sample==i),"n_read"])})*100,2),"%")
dataPlot$percent <- c(statsFile$p_u_aligned, statsFile$p_m_aligned, statsFile$p_rrna, p_unmapped)

gg <- ggplot(dataPlot, aes(x=sample, y=n_read, fill=type, text=percent)) + geom_bar(stat="identity") + scale_fill_manual(name = "", values=c("orange", "darkorange", "red", "gray")) + 
   theme_minimal() +
   labs(title = "Mapping statistics per sample", x = "Sample(s)", y = "Number of reads") + 
   theme(axis.text.x = element_text(angle = 45, hjust = 1, size=11), axis.title.x = element_text(size = 13, face="bold"), axis.title.y = element_text(size = 13, face="bold"))

ggly_b <- plotly_build(gg)
ggly_b <- inactivate_hover(ggly_b)
ggly_b <- plotly_margins(ggly_b)

ggly_b

```

<a href="#">Back to top</a>

<!--###Duplicate reads

```{r duplicateReadsPlot, fig.width=10, fig.height=4,echo=FALSE,eval=FALSE}

dataPlot <- statsFile %>%
	select(., sample, numberOfMappedReadsQCPassed, numberOfDuplicates) %>% 
	set_colnames(., c("sample", "mapped", "duplicates")) %>% 
	mutate(., noduplicates=mapped-duplicates) %>%  
	gather(., key="type", value="n_read", -sample, -mapped)

percents <- as.numeric(gsub( "%", "", statsFile$percentOfDuplicates))
percents <- 100-percents
percents <- paste(percents, "%", sep="")
percents <- c(as.character(statsFile$percentOfDuplicates), percents)
dataPlot$percents <- percents

gg <- ggplot(dataPlot, aes(x=sample, y=n_read, fill=type, text=percents)) + geom_bar(stat="identity") + 
   scale_fill_manual(name = "", values=c("gray", "orange")) + 
   theme_minimal() +   
   labs(title = "Duplicate reads per sample", x = "Sample", y = "Number of reads") + 
   theme(axis.text.x = element_text(angle = 45, hjust = 1, size=11), axis.title.x = element_text(size = 13, face="bold"), axis.title.y = element_text(size = 13, face="bold"))

ggly_b <- plotly_build(gg)
ggly_b <- plotly_margins(ggly_b)

ggly_b

```

<a href="#">Back to top</a>
-->

##Gene-based Annotation
| <font size="5">Reference: `r basename(gtf)`</font> | <font size="5"></font> |
| :------------- | -------------: |
|<font size="5">`r paste("<a href='", file.path(input_path,"report/tablecounts_raw.csv"), "'>Raw Count Table</a>")`</font> | <font size="5">`r paste0("<a href='", file.path(input_path,"report/tablecounts_tpm.csv"), "'>TPM Count Table</a>")`</font> |


```{r table3, eval=TRUE}

datatable(format(statsFile[,c('inputID', 'sampleID', 'n_u_aligned', 'n_exons', 'p_exons', 'n_introns', 'p_introns', 'n_inter', 'p_inter')], big.mark=",",scientific=FALSE), colnames =c('Sample', 'Biosample', 'Number of uniquely mapped reads', 'Number of exonic reads', 'Percent of exonic reads', 'Number of intronic reads', 'Percent of intronic reads', 'Number of intergenic reads', 'Percent of intergenic reads'), rownames = FALSE, filter = 'top', caption = 'Annotation results', options = list(columnDefs = list(list(className = 'dt-center', targets = 2:5))))

```

```{r annotPlot, fig.width=10, fig.height=7, echo=FALSE, eval=TRUE}
verboseannot <- c("Exonic reads", "Intronic reads", "Intergenic reads", "Non-annotated")
names(verboseannot) <- c("exons","introns","inter","nonannot")

dataPlot <- statsFile %>%
	select(., inputID, n_u_aligned, n_exons, n_introns, n_inter) %>% 
	set_colnames(., c("sample", "uniqueHits", "exons", "introns", "inter")) %>% 
	mutate(., nonannot=(uniqueHits - exons - introns - inter )) %>%  
  gather(., key="type", value="n_read", -sample, -uniqueHits)

dataPlot$percent <-  round(dataPlot$n_read/rep(statsFile$n_u_aligned, 4)*100, 2)


dataPlot$type=verboseannot[dataPlot$type]
dataPlot$type <- factor(dataPlot$type, levels=c("Exonic reads", "Intronic reads", "Intergenic reads", "Non-annotated"))

gg <- ggplot(dataPlot, aes(x=sample, y=n_read, fill=type, text = paste("--Sample: ",sample,"<br>","Percent: ",percent,"%<br>","n_read: ",n_read))) + 
   geom_bar(stat="identity") + scale_fill_manual(name = "", values=c("#E41A1C", "#377EB8", "#4DAF4A", "gray")) + 
   theme_minimal() +
   labs(title = "Annotation of uniquely aligned reads", x = "Sample(s)", y = "Number of reads") + 
   theme(axis.text.x = element_text(angle = 45, hjust = 1, size=11), axis.title.x = element_text(size = 13, face="bold"), axis.title.y = element_text(size = 13, face="bold"))

ggly_b <- plotly_build(gg)
for (i in 1:length(ggly_b$x$data)){
     #ggly_b$x$data[[i]]$text <- sapply(strsplit(ggly_b$x$data[[i]]$text, split="--"),"[[",2)
     ggly_b$x$data[[i]]$hoverinfo <- "none"
}
ggly_b <- inactivate_hover(ggly_b)
ggly_b <- plotly_margins(ggly_b)
ggly_b

```
<font size="2"> Reads have been annotated using known genes annotation. Reads spanning several annotation have been prioritized such as exon >> intron >> intergenic >> non-annotated.Non-annotated regions mainly correspond to DNA contigs without any gene annotations.</font>
<br>
<hr>

```{r annot, fig.width=10, fig.height=7, eval=TRUE, message=FALSE, warning=FALSE}

counts.tpm <- read.csv(file.path(input_path,"report/tablecounts_tpm.csv"), row.names=1)

## count number of expressed genes                                                       
idx <- which(rowSums(counts.tpm)>0)
counts.tpm <- counts.tpm[idx,,drop=FALSE]

## gene annotation
gtf <- rtracklayer::import(gtf)
my_genes <- gtf[gtf$type == "gene"]
mcols(my_genes) <- mcols(my_genes)[c("gene_id", "gene_type","gene_name")]
n_items <- 5
d2p <- as.matrix(as.data.frame(lapply(as.list(counts.tpm), function(x){
	    n_ex <- length(which(x>1))
	    ids <- sapply(strsplit(rownames(counts.tpm)[which(x>1)], "\\|"), "[[", 1)
	    dt <- table(factor(my_genes$gene_type[match(ids, my_genes$gene_id)], levels=unique(sort(my_genes$gene_type))))
	    c(total=n_ex, dt)
})))

## remove zero values and sort by counts
d2p <- d2p[-which(rowSums(d2p)==0),]
d2p <- d2p[order(rowSums(d2p), decreasing=TRUE),]

## reduce
if (nrow(d2p) > (n_items + 1)){
   d2p <- rbind(d2p[1:n_items,], others=colSums(d2p[(n_items+1):nrow(d2p), ]))
}

gg <- ggplot(melt(d2p), aes(x=Var2, y=value, fill=Var1)) + 
   geom_bar(stat="identity", position=position_dodge()) +
   theme_minimal() + 
   scale_fill_manual(values=c("gray", brewer.pal(n_items, "Set2"))) +
   xlab("") + labs(title="Genes annotation", x="Sample(s)", y="Expressed genes (TPM>1)") + 
   theme(legend.title=element_blank(), legend.position="bottom") +
   theme(axis.text.x = element_text(angle = 45, hjust = 1, size=11), axis.title.x = element_text(size = 13, face="bold"), axis.title.y = element_text(size = 13, face="bold"))

ggly_b <- plotly_build(gg)
ggly_b <- inactivate_hover(ggly_b)
ggly_b <- plotly_margins(ggly_b)

ggly_b

```
<br><font size="2"> Genes with a TPM value > 1 are considered as expressed. The number of expressed genes per sample was splitted into the 5 most abundant classes following the known gene annotations.</font><br>


<a href="#">Back to top</a>
